File: Graph/MostStonesRemoved/most.cpp
#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
    void countNodeInConnComp(vector<bool>& visited,
                             vector<vector<int>>& stones,
                             int node) {

        queue<int> qu;
        qu.push(node);

        int totalStones = stones.size();
        visited[node] = true;

        while (!qu.empty()) {
            int currNode = qu.front();
            qu.pop();

            for (int neighbour = 0; neighbour < totalStones; neighbour++) {

                if (!visited[neighbour]) {

                    // Same row OR same column
                    if (stones[currNode][0] == stones[neighbour][0] ||
                        stones[currNode][1] == stones[neighbour][1]) {

                        visited[neighbour] = true;
                        qu.push(neighbour);
                    }
                }
            }
        }
    }

    int removeStones(vector<vector<int>>& stones) {

        int totalNodes = stones.size();
        int totalComps = 0;

        vector<bool> visited(totalNodes, false);

        for (int i = 0; i < totalNodes; i++) {

            if (!visited[i]) {
                totalComps++;
                countNodeInConnComp(visited, stones, i);
            }
        }

        return totalNodes - totalComps;
    }
};

File: Graph/MostStonesRemoved/most.py
from collections import deque
from typing import List

class Solution:
    def countNodeInConnComp(self, visited, stones, node):
        qu = deque()
        qu.append(node)
        totalStones = len(stones)
        visited[node] = True
        while len(qu):
            currNode = qu.popleft()
            for neighbour in range(totalStones):
                if not visited[neighbour]:
                    # Same row OR same column
                    if (stones[currNode][0] == stones[neighbour][0] or
                        stones[currNode][1] == stones[neighbour][1]):

                        visited[neighbour] = True
                        qu.append(neighbour)
    def removeStones(self, stones: List[List[int]]) -> int:
        totalNodes = len(stones)
        totalComps = 0
        visited = [False] * totalNodes
        for i in range(totalNodes):
            if not visited[i]: 
                totalComps += 1
                self.countNodeInConnComp(visited,stones,i)
        return totalNodes-totalComps